hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java

```diff
@@ -0,0 +1,303 @@
+
+								collection.afterRowInsert( this, entry, i );
+							}
+					);
+					loop = loop.thenCompose( total -> reactiveConnection
+							.update( getSQLInsertRowString(), paramValues )
+							.thenApply( insertCount -> total + insertCount ) );
+				}
+			}
+		}
+
+		return loop.thenAccept( total -> LOG.debugf( "Done inserting collection: %s rows inserted", total ) );
+	}
+
+	public CompletionStage<Integer> removeReactive(Serializable id, SharedSessionContractImplementor session)
+			throws HibernateException {
+		ReactiveConnection reactiveConnection = getReactiveConnection( session );
+		if ( !isInverse && isRowDeleteEnabled() ) {
+			if ( LOG.isDebugEnabled() ) {
+				LOG.debugf(
+						"Deleting collection: %s",
+						MessageHelper.collectionInfoString( this, id, getFactory() )
+				);
+			}
+
+			List<Object> params = new ArrayList<>();
+			params.add( id );
+			String sql = getSQLDeleteString();
+			return reactiveConnection.update( sql, params.toArray( new Object[0] ) );
+		}
+		return CompletionStages.completedFuture( -1 );
+	}
+
+	@Override
+	public CompletionStage<Void> reactiveDeleteRows(
+			PersistentCollection collection,
+			Serializable id,
+			SharedSessionContractImplementor session) throws HibernateException {
+
+		if ( isInverse ) {
+			return CompletionStages.voidFuture();
+		}
+
+		if ( !isRowDeleteEnabled() ) {
+			return CompletionStages.voidFuture();
+		}
+
+		CompletionStage<Integer> loop = CompletionStages.zeroFuture();
+		ReactiveConnection reactiveConnection = getReactiveConnection( session );
+
+		boolean deleteByIndex = hasIndex && !indexContainsFormula;
+
+		Iterator deletes = collection.getDeletes( this, !deleteByIndex );
+
+		int offset = 1;
+		int i = 0;
+
+		while ( deletes.hasNext() ) {
+			final Object entry = deletes.next();
+			if ( collection.entryExists( entry, i ) ) {
+				Object[] paramValues = PreparedStatementAdaptor.bind(
+						st -> {
+							int loc = offset;
+							if ( hasIdentifier ) {
+								writeIdentifier( st, entry, loc, session );
+							}
+							else {
+								loc = writeKey( st, id, loc, session );
+								if ( deleteByIndex ) {
+									writeIndexToWhere( st, entry, loc, session );
+								}
+								else {
+									writeElementToWhere( st, entry, loc, session );
+								}
+							}
+						}
+				);
+				loop = loop.thenCompose( total -> reactiveConnection
+						.update( getSQLDeleteRowString(), paramValues )
+						.thenApply( deleteCount -> total + deleteCount ) );
+			}
+		}
+
+		return loop.thenAccept( total -> LOG.debugf( "Done deleting rows: %s deleted", total ) );
+	}
+
+	@Override
+	public CompletionStage<Void> reactiveUpdateRows(
+			PersistentCollection collection,
+			Serializable id,
+			SharedSessionContractImplementor session) throws HibernateException {
+
+		// Currently there is no "updateRows()" method in AbstractCollectionPersister
+		// Not sure if this method, reactiveUpdateRows() is needed.
```

Comment:
This comment is wrong: [BasicCollectionPersister#updateRows](https://github.com/hibernate/hibernate-orm/blob/5.4/hibernate-core/src/main/java/org/hibernate/persister/collection/AbstractCollectionPersister.java#L1746)